lib/hypervisor/hv_kvm/bus_manager.py

```diff
@@ -0,0 +1,194 @@
+
+  @abstractmethod
+  def get_next_allocation(self) -> BusAllocation:
+    """
+    Return next available bus slot address for device.
+    """
+    pass
+
+  @abstractmethod
+  def release(self, allocation: BusAllocation) -> None:
+    """
+    Releases an allocation (e.g. after HotDel).
+    """
+    pass
+
+  @abstractmethod
+  def reserve(self, allocation: BusAllocation) -> None:
+    """
+    Mark an allocation as reserved.
+    """
+    pass
+
+
+class PCIAllocator(BusAllocator):
+  _PCI_BUS = "pci.0"
+
+  def __init__(self, max_slots: int, reserved_slots: int):
+    self._max_slots = max_slots
+    self._reserved_slots = reserved_slots
+    self._occupied_slots: Set[int] = set()
+
+  @property
+  def bus_type(self) -> str:
+    return "pci"
+
+  def get_next_allocation(self) -> BusAllocation:
+    slot = self._find_free_slot()
+    return BusAllocation(
+      bus=self._PCI_BUS,
+      bus_type=self.bus_type,
+      device_params={
+        "addr": hex(slot),
+      }
+    )
+
+  def release(self, allocation: BusAllocation) -> None:
+    slot = allocation.device_params["addr"]
+    self._occupied_slots.remove(int(slot, 16))
+
+  def reserve(self, allocation: BusAllocation) -> None:
+    slot = allocation.device_params["addr"]
+    # mark slot as occupied
+    self._occupied_slots.add(int(slot, 16))
+
+  def _find_free_slot(self):
+    for slot in range(self._reserved_slots, self._max_slots):
+      if slot not in self._occupied_slots:
+        return slot
+    raise RuntimeError("No free slots available")
+
+  def initialize_from_device_info(self, device_infos: List[Dict]):
+    for device_info in device_infos:
+      if "bus" in device_info and device_info["bus"] == self._PCI_BUS:
+        slot = device_info["addr"]
+        slot = int(slot, 16)
+        self._occupied_slots.add(slot)
+
+
+class SCSIAllocator(PCIAllocator):
+  _SCSI_BUS = "scsi.0"
+
+  @property
+  def bus_type(self) -> str:
+    return "scsi"
+
+  def get_next_allocation(self) -> BusAllocation:
+    slot = self._find_free_slot()
+    return BusAllocation(
+      bus=self._SCSI_BUS,
+      bus_type=self.bus_type,
+      device_params={
+        "channel": 0,
+        "scsi-id": slot,
+        "lun": 0,
+      }
+    )
+
+  def reserve(self, allocation: BusAllocation) -> None:
+    slot = allocation.device_params["scsi-id"]
+    # mark slot as occupied
+    self._occupied_slots.add(slot)
+
+  def initialize_from_device_info(self, device_infos: List[Dict]):
+    for device_info in device_infos:
+      if "bus" in device_info and device_info["bus"] == self._SCSI_BUS:
+        slot = device_info["scsi-id"]
+        self._occupied_slots.add(slot)
+
+
+class BusAllocatorManager:
+  _DISK_TARGET = constants.HOTPLUG_TARGET_DISK
+  _SCSI_DEVICE_TYPES = constants.HT_SCSI_DEVICE_TYPES
+
+  def __init__(self, pci_allocator: PCIAllocator = None,
+               scsi_allocator: SCSIAllocator = None):
+    self._pci_allocator = pci_allocator
+    self._scsi_allocator = scsi_allocator
```

Comment:
Is it possible for BusAllocatorManager to have a list of BusAllocator instead of different concrete implementations? This currently violates the open/close principle.